File: cpp/tests/emu/emu_test.cpp
#include "../../emu/cpu.hpp"
#include "../tester.hpp"
#include <fstream>
#include <iostream>
#include <sstream>

bool emu_test_func(const std::string& test_name, const json& data) {
    CPU cpu;
    std::vector<int> program = data["program"].get<std::vector<int>>();
    cpu.load_program(program);

    std::stringstream buffer;
    std::streambuf* old = std::cout.rdbuf(buffer.rdbuf());

    // Safety cycle limit
    int cycles = data.value("max_cycles", 100);
    cpu.run(cycles);

    std::cout.rdbuf(old);
    std::string output = buffer.str();

    bool success = true;

    // Check Output
    if (data.contains("expected_output")) {
        std::string expected_output = data["expected_output"];
        if (output != expected_output) {
            std::cout << "Output mismatch in " << test_name << ": Expected " << expected_output << ", got " << output << std::endl;
            success = false;
        }
    }

    // Check PC
    if (data.contains("expected_pc")) {
        int expected_pc = data["expected_pc"];
        if (cpu.pc != expected_pc) {
            std::cout << "PC mismatch in " << test_name << ": Expected " << expected_pc << ", got " << cpu.pc << std::endl;
            success = false;
        }
    }

    // Check Memory
    if (data.contains("expected_memory")) {
        auto expected_memory = data["expected_memory"];
        for (auto it = expected_memory.begin(); it != expected_memory.end(); ++it) {
            int addr = std::stoi(it.key());
            int val = it.value();
            if ((size_t)addr >= cpu.memory.size() || cpu.memory[addr] != val) {
                std::cout << "Memory mismatch at address " << addr << " in " << test_name << ": Expected " << val;
                if ((size_t)addr < cpu.memory.size()) std::cout << ", got " << cpu.memory[addr];
                else std::cout << " (out of bounds)";
                std::cout << std::endl;
                success = false;
            }
        }
    }

    return success;
}

int main() {
    std::string data_path = "tests/emu/emu_test_data.json";
    std::ifstream file(data_path);
    if (!file.is_open()) {
        std::cerr << "Could not open data file: " << data_path << std::endl;
        return 1;
    }

    json test_data;
    file >> test_data;

    tester::run<json>("Emulator Tests", test_data, emu_test_func);

    return 0;
}

File: python/tests/emu/emu_test.py
import os
import sys
import json
import io

# Add parent directory to sys.path for imports
sys.path.append(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from emu.cpu import CPU
import tests.tester as tester

def emu_test(test_name, data):
    cpu = CPU()
    cpu.load_program(data["program"])
    
    output_buffer = io.StringIO()
    original_stdout = sys.stdout
    sys.stdout = output_buffer

    # Safety cycle limit
    cycles = data.get("max_cycles", 100)
    cpu.run(cycles)

    sys.stdout = original_stdout
    output = output_buffer.getvalue()

    success = True

    # Check Output
    if "expected_output" in data and output != data["expected_output"]:
        print(f"Output mismatch in {test_name}: Expected {repr(data['expected_output'])}, got {repr(output)}")
        success = False

    # Check PC
    if "expected_pc" in data and cpu.pc != data.get("expected_pc"):
        print(f"PC mismatch in {test_name}: Expected {data['expected_pc']}, got {cpu.pc}")
        success = False

    # Check Memory
    if "expected_memory" in data:
        for addr_str, expected_val in data["expected_memory"].items():
            addr = int(addr_str)
            if cpu.memory[addr] != expected_val:
                print(f"Memory mismatch at address {addr} in {test_name}: Expected {expected_val}, got {cpu.memory[addr]}")
                success = False

    return success

if __name__ == "__main__":
    with open(os.path.join(os.path.dirname(__file__), "emu_test_data.json"), "r") as f:
        test_data = json.load(f)
    tester.run("Emulator Tests", test_data, emu_test)
